site/content/entry/citra-progress-report-2017-q4.md

```diff
@@ -0,0 +1,202 @@
+(ab)used in some cases, such as the Home Menu Themes, in that every single theme
+that exists is actually a different CXI within the same CIA. Just tick the themes
+the user owns in the bitfield and then attach those CXIs at the end of it. Dead
+simple way for it to work perfectly with both streaming of the file, and making
+personalized CIAs for each eShop user.
+
+{{< figure src="/images/entry/citra-progress-report-2017-q4/menucia.png" 
+    title="A tiny portion of the Home Menu Themes CIA" >}}
+
+For the `am` side of streaming installations, the way it handles a request to
+install an application is by creating and then giving the requester a handle to
+a virtual file. This leaves most of the busy work to `am`, while the app only needs
+to worry about writing the CIA file into that handle.
+
+[shinyquagsire23](https://github.com/shinyquagsire23) reimplemented the entirety
+of how Citra handles NCCH files, which quickly paved the way to implementing `am`
+in Citra. This allowed users to use FBI to install  applications from its virtual
+SD card. He and [BreadFish64](https://github.com/BreadFish64) also went a step
+further and added support for installing CIAs via the SDL and Qt frontends
+directly, removing the need to use FBI. Do note that you cannot yet run installed
+applications and games directly, but CXI executables can access the virtual SD
+card for DLC, updates, etc.
+
+## [Qtifw build installer](https://github.com/citra-emu/citra/pull/2966) by [j-selby](https://github.com/j-selby)
+
+Citra now has a fancy new installer and updater, thanks to [j-selby](https://github.com/j-selby)
+and [jroweboy](https://github.com/jroweboy)'s efforts! If you haven't already, we
+strongly recommend you download and install Citra through it, since it will make
+updating as easy as re-running the installer. You can check it out over [here](/download/),
+
+## [macOS: Build x86_64h slice](https://github.com/citra-emu/citra/pull/2982) by [MerryMage](https://github.com/MerryMage)
+
+A little known feature in macOS is fat binaries, the ability to have the same executable
+contain binaries for multiple architectures. This was used extensively around 2007
+to support Mac OS X's transition from PowerPC to x86_64, allowing developers to
+have one binary work on all Macs effortlessly.
+
+After poking around Apple's LLVM fork, [MerryMage](https://github.com/MerryMage)
+found that they had a specific flag to build binaries for Intel Haswell and later.
+Enabling it made a fat binary that contained both pre-Haswell and post-Haswell code,
+allowing Citra to take advantage of the newer CPU's instructions, without dropping
+support for older ones.
+
+## [Kernel/IPC: Add a small delay after each SyncRequest to prevent thread starvation.](https://github.com/citra-emu/citra/pull/3091) by [Subv](https://github.com/Subv)
+
+When communicating with services, titles on a real Nintendo 3DS expect to wait some
```

Comment:
This isn't correct. When a specific call is made (`SyncRequest`) the 3ds puts the calling thread to sleep until the request is done. Some games rely on this behavior as a sort of priority inversion, allowing a lower priority thread a chance to run while this thread is asleep. Because citra would respond immediately, the lower priority thread wasn't getting scheduled to run. The time that we put here was the average amount of time it took to respond, but this isn't taking into account that the delay is different for different server sessions. So in the example of reading from romfs, we likely will need a different value to delay for (more re needs to be done first)

Reply:
update ds wrote it https://gist.github.com/ds84182/ecdbbd25b56a29bd4e5b32a7544b8e92